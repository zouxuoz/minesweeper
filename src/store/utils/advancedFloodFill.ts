export const advancedFloodFill = <Cell>(
  board: Array<Array<Cell>>,
  columnIndex: number,
  rowIndex: number,
  callbackFn: (cell: Cell) => Cell,
  shouldFill: (cell: Cell) => boolean,
  shouldContinue: (cell: Cell) => boolean,
): Array<Array<Cell>> => {
  const columns = board.length;
  const rows = board[0].length;

  const nextBoard = JSON.parse(JSON.stringify(board));

  const queueColumn = [columnIndex];
  const queueRow = [rowIndex];

  while (queueColumn.length > 0 && queueRow.length > 0) {
    const currentColumnIndex = queueColumn.pop() as number;
    const currentRowIndex = queueRow.pop() as number;

    let eastColumnIndex = currentColumnIndex;
    let westColumnIndex = currentColumnIndex;

    while (
      eastColumnIndex < columns - 1 &&
      shouldContinue(nextBoard[currentRowIndex][eastColumnIndex]) &&
      shouldFill(nextBoard[currentRowIndex][eastColumnIndex + 1])
    ) {
      eastColumnIndex++;
    }

    while (
      westColumnIndex >= 1 &&
      shouldContinue(nextBoard[currentRowIndex][westColumnIndex]) &&
      shouldFill(nextBoard[currentRowIndex][westColumnIndex - 1])
    ) {
      westColumnIndex--;
    }

    for (let index = westColumnIndex; index <= eastColumnIndex; index++) {
      if (
        currentRowIndex + 1 < rows &&
        shouldContinue(nextBoard[currentRowIndex][index]) &&
        shouldFill(nextBoard[currentRowIndex + 1][index])
      ) {
        queueColumn.push(index);
        queueRow.push(currentRowIndex + 1);
      }

      if (
        currentRowIndex - 1 >= 0 &&
        shouldContinue(nextBoard[currentRowIndex][index]) &&
        shouldFill(nextBoard[currentRowIndex - 1][index])
      ) {
        queueColumn.push(index);
        queueRow.push(currentRowIndex - 1);
      }
      nextBoard[currentRowIndex][index] = callbackFn(nextBoard[currentRowIndex][index]);
    }
  }

  return nextBoard;
};
