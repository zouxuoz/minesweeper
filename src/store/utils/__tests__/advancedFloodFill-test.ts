import { advancedFloodFill } from '../advancedFloodFill';

it('Should flood fill area.', () => {
  const area = [
    [0, 0, 0, 1],
    [0, 0, 1, 0],
    [0, 1, 0, 0],
    [0, 0, 0, 0],
  ];

  expect(
    advancedFloodFill(
      area,
      0,
      0,
      () => 2,
      cell => cell === 0,
      cell => cell === 0,
    ),
  ).toMatchSnapshot();
});
