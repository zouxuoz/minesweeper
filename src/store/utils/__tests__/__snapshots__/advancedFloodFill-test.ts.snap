// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should flood fill area. 1`] = `
Array [
  Array [
    2,
    2,
    2,
    1,
  ],
  Array [
    2,
    2,
    1,
    2,
  ],
  Array [
    2,
    1,
    2,
    2,
  ],
  Array [
    2,
    2,
    2,
    2,
  ],
]
`;
